File: ts/src/control-plane/contract-probes/runner.ts
/**
 * AC-728: contract-probe suite runner.
 *
 * Library-level entry point that dispatches a JSON-defined probe suite
 * across all seven AC-728 probes (directory / terminal / service / artifact
 * / cleanup / media / distributed) and aggregates results. The runner is
 * pure: callers do the IO to populate the observations; the runner verifies.
 *
 * The probes themselves stay in `./index.js`. This module adds:
 *
 * - `ContractProbeSuiteSchema` (Zod): validates the JSON spec format.
 *   Every nested object is `.strict()` so a typo like `requiredStdoutPattern`
 *   (missing `s`) fails validation rather than silently being dropped --
 *   for a JSON harness contract, malformed expectations must reject, not
 *   disappear.
 * - `runContractProbeSuite(suite)`: dispatches every invocation through the
 *   matching probe and aggregates results. The result is a discriminated
 *   union over `kind`, so each variant preserves its probe's specific
 *   failure type (e.g. `DirectoryContractFailure` carries `path`,
 *   `DistributedContractFailure` carries optional `rank` / `key`); callers
 *   switch on `kind` and access the typed fields without casts.
 * - `loadContractProbeSuite(path)`: file loader mirroring the
 *   `loadContract` pattern from `cli-contract.ts`.
 *
 * JSON wire format notes:
 *
 * - RegExp values are serialised as either a bare string (`"^trace\\."`,
 *   treated as the pattern with no flags) or as `{ source, flags? }`.
 *   The schema transforms either form to a real RegExp before handing it
 *   to the probe. Construction errors (`new RegExp` throwing
 *   `SyntaxError`) are reported via Zod issues so `safeParse` returns
 *   `{ success: false }` rather than throwing.
 * - Date values (cleanup probe's `now` and per-entry `mtime`) are
 *   serialised as ISO-8601 strings and parsed into Date objects by the
 *   schema. Malformed dates are reported via Zod issues, not raw throws.
 */

import { readFileSync } from "node:fs";

import { z } from "zod";

import {
  probeArtifactContract,
  probeCleanupContract,
  probeDirectoryContract,
  probeDistributedContract,
  probeMediaContract,
  probeServiceContract,
  probeTerminalContract,
} from "./index.js";
import type {
  ArtifactContractFailure,
  CleanupContractFailure,
  DirectoryContractFailure,
  DistributedContractFailure,
  MediaContractFailure,
  ServiceContractFailure,
  TerminalContractFailure,
} from "./index.js";

// --- Reusable Zod helpers ---------------------------------------------------

// RegExp values arrive as a bare pattern string or as a `{ source, flags? }`
// object. Both forms transform into a real RegExp via a safeParse-friendly
// transform: construction errors land as Zod issues rather than raw throws,
// so callers can rely on `safeParse` returning `{ success: false }` instead
// of catching SyntaxError separately.
const RegExpJson = z
  .union([z.string(), z.object({ source: z.string(), flags: z.string().optional() }).strict()])
  .transform((value, ctx) => {
    try {
      return typeof value === "string" ? new RegExp(value) : new RegExp(value.source, value.flags);
    } catch (err) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `invalid regular expression: ${err instanceof Error ? err.message : String(err)}`,
      });
      return z.NEVER;
    }
  });

// Date values arrive as either an existing Date or an ISO-8601 string.
// Same safeParse posture: invalid dates land as Zod issues.
const DateJson = z.union([z.string(), z.date()]).transform((value, ctx) => {
  if (value instanceof Date) {
    return value;
  }
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: `invalid ISO-8601 date: ${JSON.stringify(value)}`,
    });
    return z.NEVER;
  }
  return parsed;
});

// --- Per-probe input schemas ------------------------------------------------
//
// Every object schema is `.strict()`: unknown keys fail validation rather
// than being silently stripped. This catches typos like
// `requiredStdoutPattern` (missing `s`) at parse time -- otherwise the
// missing-`s` field would disappear and the runner would silently skip the
// expectation, returning `passed: true` for what was supposed to be a
// stronger contract.

const DirectoryInputs = z
  .object({
    presentFiles: z.array(z.string()),
    requiredFiles: z.array(z.string()),
    allowedFiles: z.array(z.string()),
    ignoredPatterns: z.array(RegExpJson).optional(),
  })
  .strict();

const TerminalInputs = z
  .object({
    exitCode: z.number(),
    stdout: z.string(),
    stderr: z.string(),
    expectedExitCode: z.number().optional(),
    requiredStdoutPatterns: z.array(RegExpJson).optional(),
    forbiddenStdoutPatterns: z.array(RegExpJson).optional(),
    requiredStderrPatterns: z.array(RegExpJson).optional(),
    forbiddenStderrPatterns: z.array(RegExpJson).optional(),
  })
  .strict();

const ServiceEndpoint = z
  .object({
    host: z.string(),
    port: z.number(),
    protocol: z.enum(["tcp", "udp"]).optional(),
  })
  .strict();

const ServiceInputs = z
  .object({
    observed: z.array(ServiceEndpoint),
    required: z.array(ServiceEndpoint),
    allowed: z.array(ServiceEndpoint).optional(),
  })
  .strict();

const ArtifactInputs = z
  .object({
    path: z.string(),
    content: z.string(),
    expectedLineEnding: z.enum(["lf", "crlf"]).optional(),
    requiredSubstrings: z.array(z.string()).optional(),
    forbiddenSubstrings: z.array(z.string()).optional(),
    requiredJsonFields: z.array(z.string()).optional(),
  })
  .strict();

const CleanupEntry = z
  .object({
    path: z.string(),
    isSymlink: z.boolean().optional(),
    symlinkTarget: z.string().optional(),
    symlinkBroken: z.boolean().optional(),
    mtime: DateJson.optional(),
  })
  .strict();

const CleanupInputs = z
  .object({
    entries: z.array(CleanupEntry),
    now: DateJson.optional(),
    maxLockfileAgeMs: z.number().optional(),
    lockfilePatterns: z.array(RegExpJson).optional(),
    sidecarPatterns: z.array(RegExpJson).optional(),
    backupPatterns: z.array(RegExpJson).optional(),
    forbidSymlinks: z.boolean().optional(),
    allowedSymlinkTargets: z.array(z.string()).optional(),
    ignoredPatterns: z.array(RegExpJson).optional(),
  })
  .strict();

const MediaInputs = z
  .object({
    path: z.string(),
    headerBytes: z.array(z.number()).optional(),
    expectedMagicBytes: z.array(z.number()).optional(),
    width: z.number().optional(),
    height: z.number().optional(),
    expectedWidth: z.number().optional(),
    expectedHeight: z.number().optional(),
    byteSize: z.number().optional(),
    minByteSize: z.number().optional(),
    maxByteSize: z.number().optional(),
    columnCount: z.number().optional(),
    expectedColumnCount: z.number().optional(),
    columnNames: z.array(z.string()).optional(),
    requiredColumnNames: z.array(z.string()).optional(),
    lineCount: z.number().optional(),
    expectedLineCount: z.number().optional(),
  })
  .strict();

const DistributedRank = z
  .object({
    rank: z.number(),
    steps: z.number().optional(),
    observations: z.record(z.string(), z.string()).optional(),
  })
  .strict();

const DistributedInputs = z
  .object({
    ranks: z.array(DistributedRank),
    worldSize: z.number().optional(),
    expectedWorldSize: z.number().optional(),
    expectedSteps: z.number().optional(),
    mustMatchAcrossRanks: z.array(z.string()).optional(),
  })
  .strict();

// --- Suite envelope ---------------------------------------------------------

export const ContractProbeKindEnum = z.enum([
  "directory",
  "terminal",
  "service",
  "artifact",
  "cleanup",
  "media",
  "distributed",
]);

export type ContractProbeKind = z.infer<typeof ContractProbeKindEnum>;

const InvocationBase = { label: z.string().optional() };

// Every invocation envelope is `.strict()` too -- unknown keys at the
// invocation level (e.g. a misspelled `inputs2`) must fail validation.
const ContractProbeInvocationSchema = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("directory"), ...InvocationBase, inputs: DirectoryInputs }).strict(),
  z.object({ kind: z.literal("terminal"), ...InvocationBase, inputs: TerminalInputs }).strict(),
  z.object({ kind: z.literal("service"), ...InvocationBase, inputs: ServiceInputs }).strict(),
  z.object({ kind: z.literal("artifact"), ...InvocationBase, inputs: ArtifactInputs }).strict(),
  z.object({ kind: z.literal("cleanup"), ...InvocationBase, inputs: CleanupInputs }).strict(),
  z.object({ kind: z.literal("media"), ...InvocationBase, inputs: MediaInputs }).strict(),
  z
    .object({ kind: z.literal("distributed"), ...InvocationBase, inputs: DistributedInputs })
    .strict(),
]);

export type ContractProbeInvocation = z.infer<typeof ContractProbeInvocationSchema>;

export const ContractProbeSuiteSchema = z
  .object({
    schema_version: z.literal(1),
    probes: z.array(ContractProbeInvocationSchema),
  })
  .strict();

export type ContractProbeSuite = z.infer<typeof ContractProbeSuiteSchema>;

// --- Results ----------------------------------------------------------------
//
// `ContractProbeRunResult` is a discriminated union over `kind` so each
// variant preserves the probe's specific failure type. TypeScript callers
// can switch on `result.kind` and access `path` / `rank` / `key` /
// `endpoint` etc. directly, without casting through a generic `unknown`.

interface RunResultBase {
  readonly label?: string;
  readonly passed: boolean;
}

export type ContractProbeRunResult =
  | (RunResultBase & {
      readonly kind: "directory";
      readonly failures: readonly DirectoryContractFailure[];
    })
  | (RunResultBase & {
      readonly kind: "terminal";
      readonly failures: readonly TerminalContractFailure[];
    })
  | (RunResultBase & {
      readonly kind: "service";
      readonly failures: readonly ServiceContractFailure[];
    })
  | (RunResultBase & {
      readonly kind: "artifact";
      readonly failures: readonly ArtifactContractFailure[];
    })
  | (RunResultBase & {
      readonly kind: "cleanup";
      readonly failures: readonly CleanupContractFailure[];
    })
  | (RunResultBase & { readonly kind: "media"; readonly failures: readonly MediaContractFailure[] })
  | (RunResultBase & {
      readonly kind: "distributed";
      readonly failures: readonly DistributedContractFailure[];
    });

/** Cross-kind failure surface for callers that iterate failures without
 * switching on `kind` (e.g. rendering a flat error report). Every probe
 * failure has at minimum `kind` and `message`. */
export interface ContractProbeFailure {
  readonly kind: string;
  readonly message: string;
}

export interface ContractProbeSuiteResult {
  readonly passed: boolean;
  readonly results: readonly ContractProbeRunResult[];
}

// --- Runner -----------------------------------------------------------------

export function runContractProbeSuite(suite: ContractProbeSuite): ContractProbeSuiteResult {
  const results: ContractProbeRunResult[] = [];

  for (const invocation of suite.probes) {
    // Dispatch is exhaustive over ContractProbeKind; the discriminated union
    // ensures that adding a new probe kind without updating this switch is a
    // compile-time error. Each branch preserves the probe's typed failure
    // shape -- no cross-kind narrowing happens here -- so callers iterating
    // `result.results` recover the full typed surface by switching on
    // `result.kind`.
    switch (invocation.kind) {
      case "directory": {
        const r = probeDirectoryContract(invocation.inputs);
        results.push({
          kind: "directory",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "terminal": {
        const r = probeTerminalContract(invocation.inputs);
        results.push({
          kind: "terminal",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "service": {
        const r = probeServiceContract(invocation.inputs);
        results.push({
          kind: "service",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "artifact": {
        const r = probeArtifactContract(invocation.inputs);
        results.push({
          kind: "artifact",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "cleanup": {
        const r = probeCleanupContract(invocation.inputs);
        results.push({
          kind: "cleanup",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "media": {
        const r = probeMediaContract(invocation.inputs);
        results.push({
          kind: "media",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
      case "distributed": {
        const r = probeDistributedContract(invocation.inputs);
        results.push({
          kind: "distributed",
          label: invocation.label,
          passed: r.passed,
          failures: r.failures,
        });
        break;
      }
    }
  }

  return {
    passed: results.every((r) => r.passed),
    results,
  };
}

export function loadContractProbeSuite(path: string): ContractProbeSuite {
  const raw = readFileSync(path, "utf-8");
  const parsed: unknown = JSON.parse(raw);
  return ContractProbeSuiteSchema.parse(parsed);
}

File: autocontext/src/autocontext/control_plane/contract_probes/runner.py
"""AC-728 contract-probe suite runner (Python parity, slice 3).

Mirrors ``ts/src/control-plane/contract-probes/runner.ts`` (PR #990).
Library-level entry point that dispatches a JSON-defined probe suite
across all seven AC-728 probes and aggregates results. Pure function:
callers do the IO to populate observations; the runner verifies.

The suite schema is intentionally separate from the probe Inputs
models (mirroring the TS Zod / interface split):

- ``ContractProbeSuiteSchema`` validates the JSON wire format with
  every nested object set to ``extra="forbid"``, so a typo like
  ``requiredStdoutPattern`` (missing ``s``) fails validation rather
  than silently being dropped.
- RegExp values can arrive as either a bare pattern string or
  ``{"source": ..., "flags": ...}``; invalid regexes surface as
  ``ValidationError``, not raw ``re.error``.
- ISO-8601 strings parse to ``datetime`` automatically.
- ``run_contract_probe_suite(suite)`` dispatches via ``kind`` and
  returns a discriminated-union list of per-probe results, each
  preserving the probe's typed failure shape.
- ``load_contract_probe_suite(path)`` is the JSON file loader.
"""

from __future__ import annotations

import json
import re
from datetime import datetime
from pathlib import Path
from typing import Annotated, Literal

from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
    StrictBool,
    StrictInt,
    model_validator,
)

from ._advanced import (
    CleanupContractFailure,
    CleanupContractProbeInputs,
    CleanupFileEntry,
    DistributedContractFailure,
    DistributedContractProbeInputs,
    DistributedRankReport,
    MediaContractFailure,
    MediaContractProbeInputs,
    probe_cleanup_contract,
    probe_distributed_contract,
    probe_media_contract,
)
from ._base import (
    ArtifactContractFailure,
    ArtifactContractProbeInputs,
    DirectoryContractFailure,
    DirectoryContractProbeInputs,
    ServiceContractFailure,
    ServiceContractProbeInputs,
    ServiceEndpointObservation,
    TerminalContractFailure,
    TerminalContractProbeInputs,
    probe_artifact_contract,
    probe_directory_contract,
    probe_service_contract,
    probe_terminal_contract,
)

__all__ = [
    "ContractProbeInvocation",
    "ContractProbeKind",
    "ContractProbeRunResult",
    "ContractProbeSuite",
    "ContractProbeSuiteResult",
    "load_contract_probe_suite",
    "run_contract_probe_suite",
]


# ---------------------------------------------------------------------------
# JSON-side helpers: RegExp + Date coercion
# ---------------------------------------------------------------------------

ContractProbeKind = Literal[
    "directory",
    "terminal",
    "service",
    "artifact",
    "cleanup",
    "media",
    "distributed",
]


def _coerce_pattern(value: object) -> re.Pattern[str]:
    """Accept a string, a dict ``{"source": ..., "flags": ...}``, or
    a pre-compiled ``re.Pattern[str]``. Invalid patterns raise
    ``ValueError`` so Pydantic surfaces them as ``ValidationError``
    rather than letting raw ``re.error`` escape.
    """
    if isinstance(value, re.Pattern):
        return value
    if isinstance(value, str):
        try:
            return re.compile(value)
        except re.error as err:
            raise ValueError(f"invalid regular expression {value!r}: {err}") from None
    if isinstance(value, dict):
        source = value.get("source")
        if not isinstance(source, str):
            raise ValueError("RegExp object missing 'source' string")
        flags_str = value.get("flags", "")
        if not isinstance(flags_str, str):
            raise ValueError("RegExp object 'flags' must be a string")
        flags = 0
        for ch in flags_str:
            mapping = {
                "i": re.IGNORECASE,
                "m": re.MULTILINE,
                "s": re.DOTALL,
                "x": re.VERBOSE,
                "u": re.UNICODE,
            }
            if ch not in mapping:
                raise ValueError(f"unknown RegExp flag {ch!r}")
            flags |= mapping[ch]
        try:
            return re.compile(source, flags)
        except re.error as err:
            raise ValueError(f"invalid regular expression {source!r}: {err}") from None
    raise ValueError(
        f"RegExp value must be a string, a {{source, flags?}} object, or a compiled pattern; got {type(value).__name__}"
    )


def _coerce_patterns(value: object) -> tuple[re.Pattern[str], ...]:
    if value is None:
        return ()
    if not isinstance(value, (list, tuple)):
        raise ValueError("expected a list of RegExp values")
    return tuple(_coerce_pattern(item) for item in value)


_PatternList = Annotated[tuple[re.Pattern[str], ...], BeforeValidator(_coerce_patterns)]


# ---------------------------------------------------------------------------
# Per-probe JSON-shape schemas
# ---------------------------------------------------------------------------
#
# Every model is `extra="forbid"`: unknown keys fail validation rather than
# being silently stripped. This catches typos like `requiredStdoutPattern`
# (missing `s`) at parse time -- otherwise the field would disappear and
# the runner would silently skip the expectation, returning passed=True
# for what was supposed to be a stronger contract.


class _Strict(BaseModel):
    """Base for every JSON-shape model.

    - ``extra="forbid"``: unknown keys reject. Mirrors TS `.strict()`.
    - ``arbitrary_types_allowed=True``: lets compiled ``re.Pattern[str]``
      ride through after the ``_coerce_patterns`` BeforeValidator.

    Primitive coercion is blocked per-field via ``StrictInt`` /
    ``StrictBool`` (PR #1006 review P3). String -> string is a no-op
    so plain ``str`` is fine; we want lists -> tuples to keep
    working, so the model is not blanket-strict.
    """

    model_config = ConfigDict(extra="forbid", arbitrary_types_allowed=True)

    @model_validator(mode="before")
    @classmethod
    def _reject_explicit_null(cls, data: object) -> object:
        # PR #1006 review P2: TS `.optional()` accepts omission but
        # not `null`. Python should reject explicit nulls so a
        # broken extractor cannot weaken declared expectations by
        # writing `requiredStdoutPatterns: null`.
        if isinstance(data, dict):
            nulls = [k for k, v in data.items() if v is None and k in cls.model_fields]
            if nulls:
                fields = ", ".join(sorted(nulls))
                raise ValueError(
                    f"explicit null not allowed for optional field(s): {fields}; omit the key to disable the expectation"
                )
        return data


class _DirectoryInputsSchema(_Strict):
    # PR #1006 review P2: TS marks these fields required (not optional).
    # A broken extractor that omits observations should fail validation,
    # not silently turn into an empty-passing suite.
    presentFiles: tuple[str, ...]
    requiredFiles: tuple[str, ...]
    allowedFiles: tuple[str, ...]
    ignoredPatterns: _PatternList | None = None

    def to_inputs(self) -> DirectoryContractProbeInputs:
        return DirectoryContractProbeInputs(
            present_files=self.presentFiles,
            required_files=self.requiredFiles,
            allowed_files=self.allowedFiles,
            ignored_patterns=self.ignoredPatterns or (),
        )


class _TerminalInputsSchema(_Strict):
    # PR #1006 review P3: use Strict* so the JSON schema rejects
    # `"exitCode": "0"` etc rather than coercing them silently.
    exitCode: StrictInt
    stdout: str
    stderr: str
    expectedExitCode: StrictInt | None = None
    requiredStdoutPatterns: _PatternList | None = None
    forbiddenStdoutPatterns: _PatternList | None = None
    requiredStderrPatterns: _PatternList | None = None
    forbiddenStderrPatterns: _PatternList | None = None

    def to_inputs(self) -> TerminalContractProbeInputs:
        return TerminalContractProbeInputs(
            exit_code=self.exitCode,
            stdout=self.stdout,
            stderr=self.stderr,
            expected_exit_code=self.expectedExitCode,
            required_stdout_patterns=self.requiredStdoutPatterns or (),
            forbidden_stdout_patterns=self.forbiddenStdoutPatterns or (),
            required_stderr_patterns=self.requiredStderrPatterns or (),
            forbidden_stderr_patterns=self.forbiddenStderrPatterns or (),
        )


class _ServiceEndpointSchema(_Strict):
    host: str
    port: StrictInt
    protocol: Literal["tcp", "udp"] | None = None

    def to_inputs(self) -> ServiceEndpointObservation:
        return ServiceEndpointObservation(host=self.host, port=self.port, protocol=self.protocol)


class _ServiceInputsSchema(_Strict):
    # PR #1006 review P2: TS marks `observed` + `required` required.
    observed: tuple[_ServiceEndpointSchema, ...]
    required: tuple[_ServiceEndpointSchema, ...]
    allowed: tuple[_ServiceEndpointSchema, ...] | None = None

    def to_inputs(self) -> ServiceContractProbeInputs:
        return ServiceContractProbeInputs(
            observed=tuple(e.to_inputs() for e in self.observed),
            required=tuple(e.to_inputs() for e in self.required),
            allowed=None if self.allowed is None else tuple(e.to_inputs() for e in self.allowed),
        )


class _ArtifactInputsSchema(_Strict):
    path: str
    content: str
    expectedLineEnding: Literal["lf", "crlf"] | None = None
    requiredSubstrings: tuple[str, ...] | None = None
    forbiddenSubstrings: tuple[str, ...] | None = None
    requiredJsonFields: tuple[str, ...] | None = None

    def to_inputs(self) -> ArtifactContractProbeInputs:
        return ArtifactContractProbeInputs(
            path=self.path,
            content=self.content,
            expected_line_ending=self.expectedLineEnding,
            required_substrings=self.requiredSubstrings or (),
            forbidden_substrings=self.forbiddenSubstrings or (),
            required_json_fields=self.requiredJsonFields or (),
        )


class _CleanupEntrySchema(_Strict):
    path: str
    isSymlink: StrictBool | None = None
    symlinkTarget: str | None = None
    symlinkBroken: StrictBool | None = None
    mtime: datetime | None = None

    def to_inputs(self) -> CleanupFileEntry:
        return CleanupFileEntry(
            path=self.path,
            is_symlink=bool(self.isSymlink),
            symlink_target=self.symlinkTarget,
            symlink_broken=bool(self.symlinkBroken),
            mtime=self.mtime,
        )


class _CleanupInputsSchema(_Strict):
    # PR #1006 review P2: TS marks `entries` required.
    entries: tuple[_CleanupEntrySchema, ...]
    now: datetime | None = None
    maxLockfileAgeMs: StrictInt | None = None
    lockfilePatterns: _PatternList | None = None
    sidecarPatterns: _PatternList | None = None
    backupPatterns: _PatternList | None = None
    forbidSymlinks: StrictBool | None = None
    allowedSymlinkTargets: tuple[str, ...] | None = None
    ignoredPatterns: _PatternList | None = None

    def to_inputs(self) -> CleanupContractProbeInputs:
        return CleanupContractProbeInputs(
            entries=tuple(e.to_inputs() for e in self.entries),
            now=self.now,
            max_lockfile_age_ms=self.maxLockfileAgeMs,
            lockfile_patterns=self.lockfilePatterns,
            sidecar_patterns=self.sidecarPatterns,
            backup_patterns=self.backupPatterns,
            forbid_symlinks=bool(self.forbidSymlinks),
            allowed_symlink_targets=self.allowedSymlinkTargets,
            ignored_patterns=self.ignoredPatterns or (),
        )


class _MediaInputsSchema(_Strict):
    path: str
    headerBytes: tuple[StrictInt, ...] | None = None
    expectedMagicBytes: tuple[StrictInt, ...] | None = None
    width: StrictInt | None = None
    height: StrictInt | None = None
    expectedWidth: StrictInt | None = None
    expectedHeight: StrictInt | None = None
    byteSize: StrictInt | None = None
    minByteSize: StrictInt | None = None
    maxByteSize: StrictInt | None = None
    columnCount: StrictInt | None = None
    expectedColumnCount: StrictInt | None = None
    columnNames: tuple[str, ...] | None = None
    requiredColumnNames: tuple[str, ...] | None = None
    lineCount: StrictInt | None = None
    expectedLineCount: StrictInt | None = None

    def to_inputs(self) -> MediaContractProbeInputs:
        return MediaContractProbeInputs(
            path=self.path,
            header_bytes=self.headerBytes,
            expected_magic_bytes=self.expectedMagicBytes,
            width=self.width,
            height=self.height,
            expected_width=self.expectedWidth,
            expected_height=self.expectedHeight,
            byte_size=self.byteSize,
            min_byte_size=self.minByteSize,
            max_byte_size=self.maxByteSize,
            column_count=self.columnCount,
            expected_column_count=self.expectedColumnCount,
            column_names=self.columnNames,
            required_column_names=self.requiredColumnNames,
            line_count=self.lineCount,
            expected_line_count=self.expectedLineCount,
        )


class _DistributedRankSchema(_Strict):
    rank: StrictInt
    steps: StrictInt | None = None
    observations: dict[str, str] | None = None

    def to_inputs(self) -> DistributedRankReport:
        return DistributedRankReport(rank=self.rank, steps=self.steps, observations=self.observations)


class _DistributedInputsSchema(_Strict):
    # PR #1006 review P2: TS marks `ranks` required.
    ranks: tuple[_DistributedRankSchema, ...]
    worldSize: StrictInt | None = None
    expectedWorldSize: StrictInt | None = None
    expectedSteps: StrictInt | None = None
    mustMatchAcrossRanks: tuple[str, ...] | None = None

    def to_inputs(self) -> DistributedContractProbeInputs:
        return DistributedContractProbeInputs(
            ranks=tuple(r.to_inputs() for r in self.ranks),
            world_size=self.worldSize,
            expected_world_size=self.expectedWorldSize,
            expected_steps=self.expectedSteps,
            must_match_across_ranks=self.mustMatchAcrossRanks,
        )


# ---------------------------------------------------------------------------
# Discriminated invocation envelope
# ---------------------------------------------------------------------------


class _DirectoryInvocation(_Strict):
    kind: Literal["directory"]
    label: str | None = None
    inputs: _DirectoryInputsSchema


class _TerminalInvocation(_Strict):
    kind: Literal["terminal"]
    label: str | None = None
    inputs: _TerminalInputsSchema


class _ServiceInvocation(_Strict):
    kind: Literal["service"]
    label: str | None = None
    inputs: _ServiceInputsSchema


class _ArtifactInvocation(_Strict):
    kind: Literal["artifact"]
    label: str | None = None
    inputs: _ArtifactInputsSchema


class _CleanupInvocation(_Strict):
    kind: Literal["cleanup"]
    label: str | None = None
    inputs: _CleanupInputsSchema


class _MediaInvocation(_Strict):
    kind: Literal["media"]
    label: str | None = None
    inputs: _MediaInputsSchema


class _DistributedInvocation(_Strict):
    kind: Literal["distributed"]
    label: str | None = None
    inputs: _DistributedInputsSchema


_InvocationUnion = (
    _DirectoryInvocation
    | _TerminalInvocation
    | _ServiceInvocation
    | _ArtifactInvocation
    | _CleanupInvocation
    | _MediaInvocation
    | _DistributedInvocation
)
ContractProbeInvocation = Annotated[_InvocationUnion, Field(discriminator="kind")]


class ContractProbeSuite(_Strict):
    """JSON wire-format suite envelope.

    Both fields are required. PR #1006 review P2: ``probes: null``
    used to coerce to an empty passing suite; the explicit-null
    rejection in ``_Strict._reject_explicit_null`` plus the lack of
    a default here makes ``null`` and omission both fail validation
    (TS rejects ``null`` the same way).
    """

    schema_version: Literal[1]
    probes: tuple[ContractProbeInvocation, ...]


ContractProbeSuiteSchema = ContractProbeSuite


# ---------------------------------------------------------------------------
# Run-result discriminated union
# ---------------------------------------------------------------------------


class _ResultBase(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid", arbitrary_types_allowed=True)
    label: str | None = None
    passed: bool


class DirectoryRunResult(_ResultBase):
    kind: Literal["directory"] = "directory"
    failures: tuple[DirectoryContractFailure, ...]


class TerminalRunResult(_ResultBase):
    kind: Literal["terminal"] = "terminal"
    failures: tuple[TerminalContractFailure, ...]


class ServiceRunResult(_ResultBase):
    kind: Literal["service"] = "service"
    failures: tuple[ServiceContractFailure, ...]


class ArtifactRunResult(_ResultBase):
    kind: Literal["artifact"] = "artifact"
    failures: tuple[ArtifactContractFailure, ...]


class CleanupRunResult(_ResultBase):
    kind: Literal["cleanup"] = "cleanup"
    failures: tuple[CleanupContractFailure, ...]


class MediaRunResult(_ResultBase):
    kind: Literal["media"] = "media"
    failures: tuple[MediaContractFailure, ...]


class DistributedRunResult(_ResultBase):
    kind: Literal["distributed"] = "distributed"
    failures: tuple[DistributedContractFailure, ...]


ContractProbeRunResult = Annotated[
    DirectoryRunResult
    | TerminalRunResult
    | ServiceRunResult
    | ArtifactRunResult
    | CleanupRunResult
    | MediaRunResult
    | DistributedRunResult,
    Field(discriminator="kind"),
]


class ContractProbeSuiteResult(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid", arbitrary_types_allowed=True)
    passed: bool
    results: tuple[ContractProbeRunResult, ...]


# ---------------------------------------------------------------------------
# Runner + loader
# ---------------------------------------------------------------------------


def run_contract_probe_suite(suite: ContractProbeSuite) -> ContractProbeSuiteResult:
    """Dispatch each invocation through the matching probe and aggregate."""
    results: list[ContractProbeRunResult] = []
    for invocation in suite.probes:
        match invocation:
            case _DirectoryInvocation():
                r = probe_directory_contract(invocation.inputs.to_inputs())
                results.append(DirectoryRunResult(label=invocation.label, passed=r.passed, failures=r.failures))
            case _TerminalInvocation():
                tr = probe_terminal_contract(invocation.inputs.to_inputs())
                results.append(TerminalRunResult(label=invocation.label, passed=tr.passed, failures=tr.failures))
            case _ServiceInvocation():
                sr = probe_service_contract(invocation.inputs.to_inputs())
                results.append(ServiceRunResult(label=invocation.label, passed=sr.passed, failures=sr.failures))
            case _ArtifactInvocation():
                ar = probe_artifact_contract(invocation.inputs.to_inputs())
                results.append(ArtifactRunResult(label=invocation.label, passed=ar.passed, failures=ar.failures))
            case _CleanupInvocation():
                cr = probe_cleanup_contract(invocation.inputs.to_inputs())
                results.append(CleanupRunResult(label=invocation.label, passed=cr.passed, failures=cr.failures))
            case _MediaInvocation():
                mr = probe_media_contract(invocation.inputs.to_inputs())
                results.append(MediaRunResult(label=invocation.label, passed=mr.passed, failures=mr.failures))
            case _DistributedInvocation():
                dr = probe_distributed_contract(invocation.inputs.to_inputs())
                results.append(DistributedRunResult(label=invocation.label, passed=dr.passed, failures=dr.failures))
    return ContractProbeSuiteResult(
        passed=all(r.passed for r in results),
        results=tuple(results),
    )


def load_contract_probe_suite(path: str | Path) -> ContractProbeSuite:
    """Load a JSON file and validate it against the suite schema."""
    raw = Path(path).read_text(encoding="utf-8")
    parsed = json.loads(raw)
    return ContractProbeSuite.model_validate(parsed)
